src/cache.rs

```diff
@@ -0,0 +1,315 @@
+use crate::access::Access;
+use crate::{CacheKey, CacheValue};
+use std::collections::{hash_map::Entry, HashMap};
+
+/// Cache entry can be in three states:
+/// - Does not exists, a given key was never inserted in the cache:             
+///     ExistsInCache::No
+/// - Exists but the value is empty.
+///      ExistsInCache::Yes(None)
+/// - Exists and contains a value:
+///     ExistsInCache::Yes(Some(value))
+pub enum ExistsInCache {
+    Yes(Option<CacheValue>),
+    No,
+}
+
+/// CacheLog keeps track of the first write and the last read for a given key.
+#[derive(Default)]
+pub struct CacheLog {
+    log: HashMap<CacheKey, Access>,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
```

Comment:
Missing doc comment:
```suggestion
An entry to be added to the cache, consisting of a `key` and an optional `value`. If the value is `None`, writing the entry will be treated as a deletion.
#[derive(PartialEq, Eq, Clone, Debug)]
```

Reply:
merged